tests/basic/pthread_detach/README.md

```diff
@@ -0,0 +1,181 @@
+    - sem down on task's scheduler semaphore
+    [[lthread sleeps and will be woken up when "task" is picked up by the LKL scheduler]]
+    - schedule_tail(abs_prev)
+```
+
+# Pthreads support
+
+## pthread_create
+The libc code calls `clone()` to create a new thread.
+There are two different types of integration points for LKL during process creation. (All of the following functions are defined in arch/lkl/kernel/threads.c):
+i) Setting up the arch specific thread_info struct.
+    - alloc_thread_stack_node()
+    - init_ti()
+    - setup_thread_stack()
+ii) Creating the host thread(lthread) backing the Linux task.
+    - copy_thread_tls()
+
+Call graph within LKL caused by pthread_create()->clone():
+```
+_do_fork(*args: kernel_clone_args)
+    - ...
+    |
+    |
+    V
+    - copy_process(...,args)
+        - ...
+        | // where is current picked from?
+        V
+        - dup_task_struct(current, node)
+            - ...
+            |
+            V
+            - alloc_thread_stack_node(node, orig) // LKL arch specific
+                - malloc thread_info.
+                |
+                V
+                - init_ti() // LKL arch specific
+                    - allocate scheduler semaphore.
+                    - set some fields to zero values: dead, prev_sched, tid, cloned_child.
+                - return pointer to thread_info.
+            |
+            V
+            - setup_thread_stack(p, orig) // LKL arch specific
+                - does no stack related setup.
+                - copies some fields from orig - flags, preempt_count, addr_limit
+        |
+        V
+        - copy_thread_tls(clone_flags, args->stack, args->stack_size,
+                             p, args->tls) // LKL arch specific
+            - if creating a thread via clone():
+                - Set flags: TIF_HOST_THREAD and TIF_CLONED_HOST_THREAD
+                - ti->tid = LKL host op for creating thread()
+                - Set process name field "task_struct->comm"
+                - Set ti.cloned_child for parent
+                - return 0 on success or -ENOMEM
+    |
+    V
+    - ...
+
+```
+
+After the clone() syscall is run, the state of newly created child process must be set as non-runnable in LKL. This in turn requires the child process be the `current` process context. If this wasn't enough there is one more complication here. For LKL to schedule the thread, it needs to transfer the CPU lock ownership to it. As mentioned earlier, CPU lock is assigned by lthread id. So briefly the lthread identity of the parent is shared with the child thread. 
+
+After setting the state as `TASK_UNINTERRUPTIBLE`, the lthread id for the child thread is restored.
+
+
+## pthread_exit
+
+A pthread issues a `SYS_exit` [here](https://github.com/lsds/sgx-lkl-musl/blob/oe_port/src/thread/pthread_create.c#L123), or if its detach attribute is set it exits [here](https://github.com/lsds/sgx-lkl-musl/blob/oe_port/src/thread/x86_64/__unmapself.s#L18).
```

Comment:
If you're including line numbers, it's probably a good idea to link to a specific revision.  This link will drift away from the anchor if the code above it is modified.